Atomic.md

```diff
@@ -0,0 +1,618 @@
+|`AtomicTxn`| |`object`|`STObject`|
+
+### 3.1. `AtomicTxn`
+
+The `AtomicTxn` inner object **must** be included in any inner transaction of an `Atomic` transaction. Its inclusion prevents hash collisions between identical transactions (since sequence numbers aren't included) and 
+
+The fields contained in this object are:
+
+| Field Name | Required? | JSON Type | Internal Type |
+|------------|-----------|-----------|---------------|
+|`Account`|✔️|`string`|`AccountID`|
+|`OuterSequence`| ✔️|`number`|`UInt32`|
+|`Sequence`| |`number`|`UInt32`|
+|`AtomicIndex`| |`number`|`UInt8`|
+
+#### 3.1.1. `Account`
+
+This is the account that is submitting the outer `Atomic` transaction.
+
+#### 3.1.2. `OuterSequence`
+
+This is the sequence number of the outer `Atomic` transaction. Its inclusion ensures that there are no hash collisions with other `Atomic` transactions.
+
+#### 3.1.3. `Sequence`
+
+This is the next available sequence number for the inner transaction's account. This only needs to be included in a multi-account `Atomic` transaction. See section 4.1 for an explanation as to why this is needed.
+
+#### 3.1.4. `AtomicIndex`
+
+This is the (0-indexed) index of the inner transaction within the existing `Atomic` transaction. The first inner transaction will have `AtomicIndex` value `0`, the second will be `1`, etc. Its inclusion ensures there are no hash collisions with other inner transactions within the same `Atomic` transaction, and that the transactions are all placed in the right order.
+
+## 4. Edge Cases of Transaction Processing
+
+### 4.1. Ledger Object ID Generation
+
+Some objects, such as [offers](https://xrpl.org/docs/references/protocol/ledger-data/ledger-entry-types/offer/#offer-id-format) and [escrows](https://xrpl.org/docs/references/protocol/ledger-data/ledger-entry-types/escrow/#escrow-id-format), use the sequence number of the creation transaction as a part of their ledger entry ID generation, to ensure uniqueness of the IDs.
+
+To get around this, in single-account `Atomic` transactions, a "phantom sequence number" will be used instead. The "phantom sequence number" will be equal to `AtomicTxn.OuterSequence + AtomicTxn.AtomicIndex`.
+
+Multi-account transactions use the same "phantom sequence number" strategy, but instead uses the equation `AtomicTxn.Sequence + AtomicTxn.AtomicIndex`, since the `OuterSequence` does not map to the inner transaction's account.
+
+### 4.2. Sequence Number Handling
+
+Section 4.1 describes how sequence numbers are used in inner transactions.
+
+The sequence numbers will always be consumed (i.e. the `AccountRoot`'s `Sequence` will be incremented) for each processed inner transaction. A transaction counts as being "processed" if it is applied to the ledger, i.e. if a `tec` or `tes` error is received. The sequence number will be incremented by the total number of inner transactions that that account had included in the `Atomic` transaction, to avoid any hash collisions.
```

Comment:
Because `AtomicIndex` is independent of which account is used on the inner transaction, I don't think it's sufficient to increment by the number of transactions for that account.

For example, A submits an `ONLYONE` type that has a mix of accounts used
```
Sequence: 10,
RawTransactions: [ 
  # For simplicity, I'm only including Account, AtomicIndex, and optional TransactionType
  A, 0,
  B, 1,
  A, 2,
  C, 3,
  A, 4,
  A, 5, OfferCreate
]
```
If #5 succeeds, then the sequence used for that offer will be `10 + 5 = 15`. If the sequence is "incremented by the total number of inner transactions that that account had included in the `Atomic` transaction", then it will only be `10 + 4 = 14`. That leaves sequence 15 available to be reused for a later `OfferCreate`, leading to a hash collision. It should end up as 16.

```suggestion
The sequence numbers will always be consumed (i.e. the `AccountRoot`'s `Sequence` will be incremented) if any inner transactions are processed. A transaction counts as being "processed" if it is applied to the ledger, i.e. if a `tec` or `tes` error is received. The sequence number for each account will be incremented by the total number of inner transactions included in the `Atomic` transaction, to avoid any hash collisions.
```
That could be specified as "The sequence number for each account that had any transactions processed", but I figure
1. Sequence numbers are cheap.
2. This allows all accounts to know definitively what their next sequence number will be if the `Atomic` succeeds, regardless of what happens inside.

Reply:
Yes, that is what I meant to say. Thanks for the note - will make it more precise.

Reply:
A thought, if we care a little bit about preserving sequence numbers - 
We could save a few sequence numbers if `UNTILFAILURE` only claims `n` sequence numbers in their processing, where `n` is the number of transactions that were processed successfully before the failure - since there is no way for a higher sequence number to have been used.

Reply:
I'm ok with that from a ledger integrity standpoint, though I do worry about the added complexity.

Reply:
That's a good point. I'll leave a note about it in the FAQ so we don't forget, but I think we can implement that later if needed, if somehow sequence numbers become more scarce.